File: pkgs/ruby/lib/opticontest/problems/salesman_euclid/random_search.rb
# frozen_string_literal: true

module Opticontest
  module Problems
    module SalesmanEuclid
      module RandomSearch
        class Config
          attr_reader :max_iterations

          def initialize(max_iterations:)
            self.max_iterations = max_iterations
          end

        private

          def max_iterations=(value)
            raise TypeError unless value.is_a? Integer
            raise unless value.positive?

            @max_iterations = value
          end
        end

        class Algorithm
          include Opticontest::Problems::SalesmanEuclid::Common

          attr_reader :config, :points, :solution

          def initialize(config, points)
            raise TypeError unless config.is_a?(Config) && points.is_a?(Points)

            @config = config
            @points = points

            @solution = solve
          end

        private

          def solve
            best_solution = Solution.random points

            config.max_iterations.times do
              current_solution = Solution.random points

              next unless current_solution.length < best_solution.length

              best_solution = current_solution
            end

            best_solution
          end
        end
      end
    end
  end
end

File: pkgs/python/opticontest/problems/salesman_euclid/random_search.py
from opticontest.problems.salesman_euclid.common import Solution

class Config:
    def __init__(self, **kwargs):
        self.__set_max_iterations(kwargs[':max_iterations'])

    def __set_max_iterations(self, value):
        value = int(value)
        if value < 1:
            raise RuntimeError('invalid value')
        self.max_iterations = value

class Algorithm:
    def __init__(self, config, points):
        self.config = config
        self.points = points

        self.solution = self.__solve()

    def __solve(self):
        best_solution = Solution.random(self.points)
        for x in range(0, self.config.max_iterations):
            current_solution = Solution.random(self.points)
            if current_solution.length < best_solution.length:
                best_solution = current_solution
        return best_solution
